Mathlib/Tactic/Pushneg.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2022 Alice Laroche. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alice Laroche
+-/
+
+import Lean
+import Lean.Meta
+import Lean.Elab
+import Mathlib.Lean.Expr
+import Mathlib.Logic.Basic
+
+namespace Mathlib.Tactic
+
+open Lean Meta Elab.Tactic Parser.Tactic
+open Lean.Expr
+open Lean.Elab.Term
+
+theorem not_not_eq (h : p = p') : (¬¬p) = p' := h ▸ (propext not_not)
+theorem not_and_eq (h : p = p')    (h' : (¬q) = q') : (¬ (p ∧ q)) = (p' → q') := h ▸ h' ▸ propext not_and
+theorem not_or_eq  (h : (¬p) = p') (h' : (¬q) = q') : (¬ (p ∨ q)) = (p' ∧ q') := h ▸ h' ▸ propext not_or_distrib
+theorem not_forall_eq {s s' : α → Prop} (h : ∀x, (¬s x) = s' x) : (¬ ∀x, s x) = (∃ x, s' x) := funext h ▸ propext not_forall
+theorem not_exists_eq {s s' : α → Prop} (h : ∀x, (¬s x) = s' x) : (¬ ∃x, s x) = (∀ x, s' x) := forall_congr h ▸ propext not_exists
+
+def binderTelescope1 (e : Expr) (f : Expr → Expr → MetaM α) : MetaM α := do
+match e with
+| Expr.lam n t b d => withLocalDecl n d.binderInfo t fun x => f x (instantiate1 b x)
+| Expr.forallE n t b d => withLocalDecl n d.binderInfo t fun x => f x (instantiate1 b x)
+| _ => throwError "binder expected{indentExpr e}"
+
+partial def pushNegation (expr : Expr) : MetaM (Expr × Expr) := do
```

Comment:
```suggestion
partial def pushNegation (expr : Expr) : MetaM Simp.Result := do
```

There are functions that chain simp results, then you don't need to bother with `isAppOf ``Eq.refl`

Reply:
Maybe I didn't find the helpers, but to me if i use Simp.Result i will need to to `match r.proof? with` instead of `isAppOf ``Eq.refl` which is pretty much the same.
And almost everywhere else i will need to use r.getProof, so it don't really seem worth it.